clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java

```diff
@@ -77,19 +75,20 @@ public static NodeApiVersions create(Collection<ApiVersion> overrides) {
     }
 
     public NodeApiVersions(Collection<ApiVersion> nodeApiVersions) {
-        this.nodeApiVersions = nodeApiVersions;
         for (ApiVersion nodeApiVersion : nodeApiVersions) {
-            // Newer brokers may support ApiKeys we don't know about, ignore them
             if (ApiKeys.hasId(nodeApiVersion.apiKey)) {
                 ApiKeys nodeApiKey = ApiKeys.forId(nodeApiVersion.apiKey);
                 short v = Utils.min(nodeApiKey.latestVersion(), nodeApiVersion.maxVersion);
                 if (v < nodeApiVersion.minVersion) {
-                    usableVersions.put(nodeApiKey, NODE_TOO_NEW);
+                    usableVersions.put(nodeApiKey, UsableVersion.tooNew(nodeApiVersion));
                 } else if (v < nodeApiKey.oldestVersion()) {
-                    usableVersions.put(nodeApiKey, NODE_TOO_OLD);
+                    usableVersions.put(nodeApiKey, UsableVersion.tooOld(nodeApiVersion));
                 } else {
-                    usableVersions.put(nodeApiKey, v);
+                    usableVersions.put(nodeApiKey, new UsableVersion(nodeApiVersion, v));
```

Comment:
Maybe we could move the logic that creates the appropriate `UsableVersion` into a static factory method (or constructor) in `UsableVersion`.